### PR TITLE
WT-11336 Check chunk cache pinned list when admitting chunks

### PR DESCRIPTION
**Pull request by Siddhartha8899**
The following work has been done in the ticket:

1. A new function was created to handle the pinned configuration due to the increasing size of the setup function and the potential needed to reconfigure this specific config in the future.
2. The pinned objects were extracted from the config and sorted into an array. Additionally, the name of each object was changed (example: from `table:abcd` to` abcd.wt`), as this is the required format for storing object names.
3. With a sorted array of pinned objects in place, a binary search is now performed every time a chunk is requested. This search determines whether the chunk belongs to any of the pinned objects.
4. Added unit testing for binary search.

### CODE REVIEW

File: src/block_cache/block_chunkcache.c
Comment (Siddhartha8899) [inline src/block_cache/block_chunkcache.c:453]: Since @korteland is changing something related to object name, I will wait for that PR to be merged and incorporate the necessary changes.

Comment (korteland) [reply src/block_cache/block_chunkcache.c:453]: No need for that - we can merge this first if it's ready first.

File: src/include/block_chunkcache.h
Comment (korteland) [inline src/include/block_chunkcache.h:46]: Nitpick, indentation seems messed up here

File: src/include/misc.h
Comment (korteland) [inline src/include/misc.h:241]: Nitpick: "an string key" -> "a string key"

File: src/include/misc.h
Comment (korteland) [inline src/include/misc.h:258]: This could do with some unit tests.

File: src/block_cache/block_chunkcache.c
Comment (korteland) [inline src/block_cache/block_chunkcache.c:282]: Nice work getting this into a different function, that's a good simplification. Is the `s` in `gets` shorted for "sorted"?

File: src/block_cache/block_chunkcache.c
Comment (Siddhartha8899) [reply src/block_cache/block_chunkcache.c:259]: 1. [This](https://github.com/wiredtiger/wiredtiger/pull/1821#issuecomment-86302940) should explain the reason.
2. The `void *` here is the address to the `char *`, therefore, I had to first typecast it to `char**` and then dereference it. Directly dereferencing a `void *` will result in compilation errors. 

> Can we just pass `strcmp` to the qsort call?

For the same reason above, I believe we cannot directly pass `strcmp` to `qsort`.

3. I picked this comment pattern from different places using Qsort. For example -

```
/*
 * __evict_lru_cmp --
 *     Qsort function: sort the eviction array.
 */
static int WT_CDECL
__evict_lru_cmp(const void *a_arg, const void *b_arg)
{
```

File: src/include/block_chunkcache.h
Comment (Siddhartha8899) [reply src/include/block_chunkcache.h:46]: The formatting is usually handled by `s_all`. After I fix it, `s_all` chooses to indent it. I have seen it in different places, for example, [here](https://github.com/wiredtiger/wiredtiger/blob/develop/src/include/btmem.h#L169).

File: src/block_cache/block_chunkcache.c
Comment (Siddhartha8899) [reply src/block_cache/block_chunkcache.c:282]: Not really, I was just following the convention of `__wt_config_gets`. Let me know if you want me to remove the s.

File: src/include/misc.h
Comment (Siddhartha8899) [reply src/include/misc.h]: Updated in [a03e37b](https://github.com/wiredtiger/wiredtiger/pull/9438/commits/a03e37bb9c24cb2624476e1dd284dfc89a52c254)

File: src/include/misc.h
Comment (Siddhartha8899) [reply src/include/misc.h:258]: I have done quite some testing on this. Let me know if you have anything specific in mind.

**Review by ruby-oujo — APPROVED** [review]

File: src/block_cache/block_chunkcache.c
Comment (korteland) [reply src/block_cache/block_chunkcache.c:259]: Hmm I still don't get the `WT_CDECL` call thing sorry. Why do we need it on this strcmp function and not anywhere else? What happens if it's not there?

For the cast and `strcmp` questions - those answers make sense, thanks for that.

File: src/include/block_chunkcache.h
Comment (korteland) [reply src/include/block_chunkcache.h:46]: Fair enough - I'll resolve this.

File: src/include/misc.h
Comment (korteland) [reply src/include/misc.h:258]: Yes, I think some unit tests would be appropriate here - something similar in spirit to the `qsort`/`qsort_r` tests. I can help get that set up if you're not familiar with them.

File: src/block_cache/block_chunkcache.c
Comment (korteland) [reply src/block_cache/block_chunkcache.c:282]: Ah I see. I believe that function is modelled after the C standard library function `gets` where `gets` is short for `get string`. In that case, I think this can simply be called `__config_get_pinned_objects`, or maybe something to show that it's sorted? e.g. `__config_sort_pinned_objects`. Not sure what the best option has, I'm happy to talk about it offline.

File: src/block_cache/block_chunkcache.c
Comment (korteland) [reply src/block_cache/block_chunkcache.c:303]: I thought I left a review comment here but I guess not - I think this is fine, the size seems to be more of a hint to avoid reallocations. Loads of other places use 0.

File: src/block_cache/block_chunkcache.c
Comment (Siddhartha8899) [reply src/block_cache/block_chunkcache.c:259]: `__cdecl` specifies a certain "calling convention".

The `WT_CDECL` macro is set up in two different files - o`s_windows.h` and `posix.h`. In the `os_windows.h` file, it's set as `__cdecl`, which is a rule that tells the compiler how to handle function calls. This rule is commonly used in Windows systems (Microsoft compiler).

The standard C function (qsort) have a compare function like this signature:

`int ( __cdecl *compare ) (const void *key, const void *datum)`

On the other hand, in the `posix.h` file, `WT_CDECL` isn't assigned anything. This is because, in `POSIX` systems don't need to specify this rule. The system's compiler automatically takes care of the details for function calls.

Comment (Siddhartha8899) [reply src/block_cache/block_chunkcache.c:259]: You would see it being used for all the compare functions in our code.

File: src/include/misc.h
Comment (Siddhartha8899) [reply src/include/misc.h:258]: That would be great!

File: src/block_cache/block_chunkcache.c
Comment (Siddhartha8899) [reply src/block_cache/block_chunkcache.c:282]: Updated in [dccdecc](https://github.com/wiredtiger/wiredtiger/pull/9438/commits/dccdecc42fb0241ad29f258f9e1448831ba61d7b)

File: src/block_cache/block_chunkcache.c
Comment (korteland) [reply src/block_cache/block_chunkcache.c:259]: OK I did a little digging and I think I see what's going on here - we need this for all C standard library functions we provide callbacks to, so `qsort` and `bsearch`. I'm OK with leaving it in - we theoretically no longer need it now that we've got our own `qsort` implementation, but `bsearch` still needs it and there's no harm leaving it.

File: src/include/misc.h
Comment (Siddhartha8899) [reply src/include/misc.h:258]: Updated in [d3a6198](https://github.com/wiredtiger/wiredtiger/pull/9438/commits/d3a619866111ee59400804b1422c2f6bb32fbdcf)

File: test/unittest/tests/test_binary_search_string.cpp
Comment (korteland) [inline test/unittest/tests/test_binary_search_string.cpp:100]: These are pretty good - I'd suggest a few more cases, e.g.
* String prefixes matching, e.g. `array = {"apple", ...}` and search for `apples`
* The other way around - have `apples` in the array and search for `apple`
* Empty strings. Is `""` in the normal array of fruit? What about if you have `array = {"", "apple", ...}`?
* Vice versa - if `array = {"", "", ...}`, is the empty string found? What about NULL?

Comment (Siddhartha8899) [reply test/unittest/tests/test_binary_search_string.cpp:100]: Nice one, will add those.

Comment (Siddhartha8899) [reply test/unittest/tests/test_binary_search_string.cpp:100]: @korteland I have added the test scenarios that you requested.

Also, the compiler starts to complain if the key is NULL. Hence, we should be good on that end. Also, binary search is unsupported if there is a NULL value present in an array. I have added a comment. 

The above is updated in [d68675c](https://github.com/wiredtiger/wiredtiger/pull/9438/commits/d68675c7854985f346b86de70f485ec92c4b5589)

**Review by korteland — APPROVED** [review]
LGTM! Good stuff.

**Pull request merged**
